repair/repair.cc

```diff
@@ -2267,9 +2268,64 @@ static std::unordered_set<locator::host_id> get_token_owners_in_dcs(std::vector<
     return dc_endpoints;
 }
 
+future<> repair::remote_metas::for_each_local_meta(std::function<future<>(const tablet_repair_task_meta&)> func) const {
+    size_t this_shard = this_shard_id();
+    if (!_metas_on_shards[this_shard]) {
+        co_return;
+    }
+    for (size_t i = 0; i < _metas_on_shards[this_shard]->metas.size(); ++i) {
+        co_await func(_metas_on_shards[this_shard]->metas[i]);
+    }
+}
+
+// func calls on different shards are synchronous.
+future<> repair::remote_metas::for_each_meta_on_all_shards(std::function<future<>(const tablet_repair_task_meta&)> func) const {
+    for (size_t shard = 0; shard < smp::count; ++shard) {
+        co_await smp::submit_to(shard, [func, this]() mutable {
+            return for_each_local_meta(func);
+        });
+    }
+}
+
+size_t repair::remote_metas::size() const {
+    return _metas_count;
+}
+
+void repair::remote_metas::clear() {
+    _metas_on_shards.clear();
+}
+
+future<> repair::remote_metas_builder::allocate_on_shard(size_t shard_id) {
+    _remote_metas._metas_on_shards[shard_id] = co_await smp::submit_to(shard_id, []() -> future<remote_metas::remote_data_ptr> {
+        auto ptr = make_lw_shared<remote_data>();
+        co_return remote_metas::remote_data_ptr(std::move(ptr));
+    });
+}
+
+future<> repair::remote_metas_builder::allocate_all_shards() {
+    return parallel_for_each(smp::all_cpus(), [this](size_t shard) -> future<> {
+        return allocate_on_shard(shard);
+    });
+}
+
+future<> repair::remote_metas_builder::add_on_shard(size_t shard_id, tablet_repair_task_meta meta) {
+    ++_remote_metas._metas_count;
+    co_await smp::submit_to(shard_id, [meta = std::move(meta), &metas_ptr = _remote_metas._metas_on_shards[shard_id], shard_id]() {
+        if (!metas_ptr) {
+            on_internal_error(rlogger, format("remote_metas_builder::add_on_shard: metas vector isn't created on shard_id={}", shard_id));
+        }
+        metas_ptr->metas.push_back(meta);
+    });
+}
+
+repair::remote_metas repair::remote_metas_builder::build() && {
+    return std::move(_remote_metas);
+}
+
 // Repair all tablets belong to this node for the given table
 future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_name, std::vector<sstring> table_names, bool primary_replica_only, dht::token_range_vector ranges_specified, std::vector<sstring> data_centers, std::unordered_set<locator::host_id> hosts, std::unordered_set<locator::host_id> ignore_nodes, std::optional<int> ranges_parallelism) {
-    std::vector<tablet_repair_task_meta> task_metas;
+    repair::remote_metas_builder task_metas_builder;
+    co_await task_metas_builder.allocate_all_shards();
```

Comment:
I think we can allocate lazily. It is useful for the case that  a large node with lots of shards. When a user calls add_on_shard, we allocate on demand. This also makes it is easier to use task_metas_builder.